core/shared/src/main/scala/fs2/concurrent/Signal.scala

```diff
@@ -48,6 +48,69 @@ trait Signal[F[_], A] {
   /** Asynchronously gets the current value of this `Signal`.
     */
   def get: F[A]
+
+  /** Returns when the condition becomes true, semantically blocking
+    * in the meantime.
+    *
+    * This method is particularly useful to transform naive, recursive
+    * polling algorithms on the content of a `Signal`/ `SignallingRef`
+    * into semantically blocking ones. For example, here's how to
+    * encode a very simple cache with expiry, pay attention to the
+    * definition of `view`:
+    *
+    *  {{{
+    * import cats.effect._
+    * import cats.syntax.all._
+    * import fs2.concurrent.SignallingRef
+    * import scala.concurrent.duration._
+    *
+    * trait Refresh[F[_], A] {
+    *   def get: F[A]
+    * }
+    * object Refresh {
+    *   def create[F[_]: Temporal, A](
+    *     action: F[A],
+    *     refreshAfter: A => FiniteDuration,
+    *     defaultExpiry: FiniteDuration
+    *   ): Resource[IO, Refresh[A]] =
+    *     Resource
+    *       .eval(SignallingRef[F, Option[Either[Throwable, A]]](None))
+    *       .flatMap { state =>
+    *         def refresh: F[Unit] =
+    *           state.set(None) >> action.attempt.flatMap { res =>
+    *             val t = res.map(refreshAfter).getOrElse(defaultExpiry)
+    *             state.set(res.some) >> Temporal[F].sleep(t) >> refresh
+    *           }
+    *
+    *         def view = new Refresh[A] {
+    *           def get: F[A] = state.get.flatMap {
+    *             case Some(res) => F.fromEither(res)
+    *             case None => state.waitUntil(_.isDefined) >> get
+    *           }
+    *         }
+    *
+    *         refresh.background.as(view)
+    *       }
+    * }
+    * }}}
+    *
+    * Note that because `Signal` prioritizes the latest update when
+    * its state is updating very quickly, completion of the `F[Unit]`
+    * might not trigger if the condition becomes true and then false
+    * immediately after.
+    *
+    * Therefore, natural use cases of `waitUntil` tend to fall into
+    * two categories:
+    * - Scenarios where conditions don't change instantly, such as
+    *   periodic timed processes updating the `Signal`/`SignallingRef`.
+    * - Scenarios where conditions might change instantly, but the `p`
+    *   predicate is monotonic, i.e. if it tests true for an event, it
+    *   will test true for the following events as well.
+    *   Examples include waiting for a unique ID stored in a `Signal`
+    *   to change, or waiting for the value of the `Signal` of an
+    *   ordered `Stream[IO, Int]` to be greater than a certain number.
+    */
+  def waitUntil(p: A => Boolean): F[Unit]
```

Comment:
Since `Signal` isn't sealed isn't adding this method a breaking change? :(

Reply:
Yeah...that sucks

Reply:
I'll add it to SignalOps and that should fix it I think

Reply:
Does it need to go on `SignalOps`? You could just put that default implementation and the `Concurrent` constraint directly here.

Reply:
I could yeah, I guess I just went with how `map` is implemented. @mpilquist opinions?

Reply:
No strong ones

Reply:
I found a clear justification for having it as an instance method: scaladoc friendliness. I think `map` has a similar problem, but then you'd have the question of whether removing `SignalOps` entirely constitutes source breakage or not, so I'll punt on `map` for now

Reply:
We can keep `SignalOps` deprecated, just remove the `implicit` modifier. So there's no source-breakage but the `map` on `Signal` will clearly take precedence.

Reply:
I'll leave that for a potential future PR